### PR TITLE
Improve pip_smoke_test.sh

### PR DESCRIPTION
**Pull request by caisq**
1) Refactor the test HTTP request into a bash funciton for modularity.
2) Previously, the script tested only the "/data/logdir" path, which
   means the test wouldn't have caught the breakage due to, e.g.,
   'webfile.zips' being omitted from package_data in setup.py, which
   has happened before. This change adds a check of the "/" path in
   order to cover that.
3) Do tensorboard process shutdown and temp file cleanup even if the
   test fails.

### CODE REVIEW

**Comment by caisq** [comment]
cc jart@, wchargin@

**Comment by caisq** [comment]
cc @jart @wchargin

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 1]
```diff
@@ -131,32 +131,51 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
-    echo "Request to ${TEST_URL} succeeded (200)!"
-    echo
-    REQUEST_SUCCEEDED=1
-    break
-  else
-    : $(( RETRY_COUNTER++ ))
-    echo "Request to ${TEST_URL} failed. Will retry in 1 second..."
+echo
+
+function test_access_url(){
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: Prefer `test_access_url() {` (no `function`). (`function` is a bash-ism; google-style permits you to use it, but only consistently, and it is not used elsewhere in this file (and omitting it is better than including it))

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 2]
```diff
@@ -131,32 +131,51 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
+  TEST_URL="$1"
+
+  echo
+  echo "Sending test HTTP requests at URL: ${TEST_URL} (${NUM_RETRIES} retries)"
+  echo
+
+  RETRY_COUNTER=0
+  REQUEST_SUCCEEDED=0
+  while [[ "${RETRY_COUNTER}" -lt "${NUM_RETRIES}" ]]; do
+    sleep 1
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: Now that this behavior is in a function, it makes more sense to `sleep 1` only in the failure case, after the "will retry in 1 second" message. As written, if you have ten URLs to query, then you'll spend a minimum of ten seconds, even if it only takes one second for the server to initialize.

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 3]
```diff
@@ -131,32 +131,51 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
-    echo
-    REQUEST_SUCCEEDED=1
-    break
-  else
-    : $(( RETRY_COUNTER++ ))
-    echo "Request to ${TEST_URL} failed. Will retry in 1 second..."
+echo
+
+function test_access_url(){
+  # Test requests to a URL with curl.
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: How about something more descriptive, like "Attempt to fetch the given URL repeatedly until an HTTP 200 response is returned, failing if this does not occur within $NUM_RETRIES requests separated by a one-second delay."?

(Google-style also wants you to document which globals are used and modified, declaring other variables as `local`, so: `local test_url="{$1}"`, etc.)

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 4]
```diff
@@ -131,32 +131,51 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
+  RETRY_COUNTER=0
+  REQUEST_SUCCEEDED=0
+  while [[ "${RETRY_COUNTER}" -lt "${NUM_RETRIES}" ]]; do
+    sleep 1
+    STATUS_CODE=$(curl -Is "${TEST_URL}" | head -1 | cut -d ' ' -f 2)
+    if [[ "${STATUS_CODE}" == 200 ]]; then
+      echo
+      echo "Request to ${TEST_URL} succeeded (200)!"
+      echo
+      REQUEST_SUCCEEDED=1
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: Can simplify this state by just `return`ing here, removing this variable, and dissolving the if-statement on line 168.

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 5]
```diff
@@ -170,5 +189,9 @@ rm -rf "${PIP_TMP_DIR}"
 rm -rf "${TMP_LOGDIR}"
 
 echo
-echo "=== Smoke test of tensorboard PIP package PASSED ==="
+if [[ "${TEST_URL_FAILED}" == 0 ]]; then
+  echo "=== Smoke test of tensorboard PIP package PASSED ==="
+else
+  echo "=== Smoke test of tensorboard PIP package FAILED ==="
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: Whoa, you want to `exit 1` here, right? (Including your final `echo` on 197, presumably.)

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 6]
```diff
@@ -170,5 +189,9 @@ rm -rf "${PIP_TMP_DIR}"
 rm -rf "${TMP_LOGDIR}"
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: I should have mentioned this in the original PR, but—is the `-f` needed here? We should prefer to use `-r` alone if possible.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 1 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: Done.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 3 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: Done.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 2 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: Done.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 4 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: Done.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 6 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: You are right: `rm -r` suffices. Done.

File: tensorboard/pip_package/pip_smoke_test.sh (hunk 5 above)
Comment (caisq) [inline tensorboard/pip_package/pip_smoke_test.sh]: Good catch. Using `die`.

**Comment by caisq** [comment]
@wchargin Also, I agree that keeping the temporary folders on failure makes debugging easier. Done. But the tensorboard subprocess is still killed on failure, because otherwise the port usage will linger.

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 7]
```diff
@@ -131,44 +131,65 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
+  #
+  # Retrying occur with a 1-second delay.
+  #
+  # Global variables used: ${NUM_RETIRES}.
+  #
+  # Usage:
+  #   test_access_url <URL>
+  # E.g.,
+  #   test_access_url http://localhost:6006/
+  local TEST_URL="$1"
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]: Local variables are `lower_camel_case` (also for others in this function).

File: tensorboard/pip_package/pip_smoke_test.sh [hunk 8]
```diff
@@ -131,44 +131,65 @@ TMP_LOGDIR=$(mktemp -d --suffix _tensorboard_logdir)
...
+
+  echo
+  echo "Sending test HTTP requests at URL: ${TEST_URL} (${NUM_RETRIES} retries)"
+  echo
+
+  local RETRY_COUNTER=0
+  while [[ "${RETRY_COUNTER}" -lt "${NUM_RETRIES}" ]]; do
+    STATUS_CODE=$(curl -Is "${TEST_URL}" | head -1 | cut -d ' ' -f 2)
```
Comment (wchargin) [inline tensorboard/pip_package/pip_smoke_test.sh]:
```sh
local status_code
status_code="$(curl …)"
```

**Comment by caisq** [comment]
@wchargin. Thanks. Done.

**Comment by caisq** [comment]
Gentle ping? @chihuahua @jart

**Pull request merged**